lms/djangoapps/instructor_task/tasks_helper/grades.py

```diff
@@ -550,8 +554,54 @@ def _graded_scorable_blocks_to_header(cls, course):
 
 
 class ProblemResponses(object):
+
+    @classmethod
+    def _build_problem_list(cls, course_blocks, root, path=None):
+        """
+        Generate a tuple of display names, block location paths and block keys
+        for all problem blocks under the ``root`` block.
+        """
+        if path is None:
+            path = [course_blocks.get_xblock_field(root, 'display_name')]
+        for block in course_blocks.get_children(root):
+            display_name = course_blocks.get_xblock_field(block, 'display_name')
+            if block.block_type == 'problem':
+                yield display_name, path, block
+            else:
+                for result in cls._build_problem_list(course_blocks, block,
+                                                      path + [display_name]):
+                    yield result
+
     @classmethod
-    def generate(cls, _xmodule_instance_args, _entry_id, course_id, task_input, action_name):
+    def _build_student_data(cls, user_id, course_id, problem_location):
+        """
+        Generate a list of problem responses for all problem under the
+        ``problem_location`` root.
+        """
+        problem_key = UsageKey.from_string(problem_location).map_into_course(course_id)
+        user = get_user_model().objects.get(pk=user_id)
+        course_blocks = get_course_blocks(user, problem_key)
```

Comment:
Is there any way to call this method with `user=None` or `user=AnonymousUser`, so that it will return the full course tree, and not a user-specific tree? That would be better. The full course tree includes the content that all cohorts would see and returns all possible children of blocks like A-B Test blocks and Randomized Content Blocks. passing in any real user returns only the specific subset of the blocks in the course that that user would see.

The answer may be "no" here, but I seem to recall there was some way to get the full course through this or a similar API. I don't think it's worth spending too much time on though.

Reply:
A user of None is not possible, but it is possible to use AnonymousUser, but that will not return the full course tree from what I understand of the code. The data returned will be based on the [user's access level](https://github.com/open-craft/edx-platform/blob/aa6f88382d46af43b5523b7980b26ebad742a553/lms/djangoapps/course_blocks/api.py#L24-L85).  Won't anyone who can use these tools will have full access to the course structure?

I think it's possible to get the full tree anyway, but my thought was that restricting it based on user permissions would be a good thing, so didn't look into it much.

Reply:
Ok, don't worry about it. We're not taking any functionality away here anyways.